### ISSUE TITLE
Why entities, list, and singles?

### ISSUE THREAD

**Issue by adam-lynch** [JiriChara/vuex-crud#5]
I'm new to Vuex. What's the reasoning behind having `entities`, `list`, and `singles` in the state and how do they differ? Why not a single list?

**Comment by JiriChara** [comment]
* `entities` is an object with all resources that have been fetched. Each key in the `entities` object represents the ID of certain resource. I prefer to have flatten object like this, because it will allow me to cache resources better and to handle nested resources in nicer way
* `list` is list of IDs of resources that are currently shown in the list. Imagine that you have a blog with 7 articles and you would like to show 5 articles per page. So if user goes to the first page then `list` object will contain IDs of first 5 articles, but when she goes to the second page, then `list` will contain only remaining two IDs. If you have endless scrolling then you can override `fetchListSuccess` mutation or change `onFetchListSuccess` callback to append the articles to your custom array.
* `singles` is array of IDs of resources that user accessed directly. For example if user goes to `/articles/1` then article will appear in the `entities` object, but it will not affect the current `list` of articles.

**Comment by adam-lynch** [comment]
>- entities is an object with all resources that have been fetched. Each key in the entities object represents the ID of certain resource. I prefer to have flatten object like this, because it will allow me to cache resources better and to handle nested resources in nicer way

Sounds good, I've read similar things elsewhere.

>- list is list of IDs of resources that are currently shown in the list. Imagine that you have a blog with 7 articles and you would like to show 5 articles per page. So if user goes to the first page then list object will contain IDs of first 5 articles, but when she goes to the second page, then list will contain only remaining two IDs. If you have endless scrolling then you can override fetchListSuccess mutation or change onFetchListSuccess callback to append the articles to your custom array.

OK good to know. I'll only have infinite scrolling.

>- singles is array of IDs of resources that user accessed directly. For example if user goes to /articles/1 then article will appear in the entities object, but it will not affect the current list of articles.

So are you saying that `singles` will be empty until `byId` is called, then one ID will be added to `singles`? Why is this useful? Why isn't `entities` enough?

**Comment by JiriChara** [comment]
Hey @adam-lynch,

sorry for late reply!

`singles` array will be empty until you call `fetchSingle` action - which goes then to `fetchSingleSuccess` mutation and add the add to the array. It can be useful when you want to display 2 articles on one page and fetch them one-by-one and also it's better to use arrays in Vue.js, because of it's reactivity system.

`findById` is a getter method that is using `singles` object internally. I have experienced some reactivity issues when I have used `entities` object directly.
